### ISSUE TITLE
Why do `MarshalTypeNearestDelta2` and `MarshalTypeNearestDelta` need validation?

### ISSUE THREAD

**Issue by KofClubs** [VictoriaMetrics/VictoriaMetrics#5692]
### Describe the bug

I noticed that the comment of this function mentioned that when merging blocks, `MarshalTypeNearestDelta2` and `MarshalTypeNearestDelta` need validation:
```go
// NeedsValidation returns true if mt may need additional validation for silent data corruption.
func (mt MarshalType) NeedsValidation() bool {
	switch mt {
	case MarshalTypeNearestDelta2,
		MarshalTypeNearestDelta:
		return true
	default:
		// Other types do not need additional validation,
		// since they either already contain checksums (e.g. compressed data)
		// or they are trivial and cannot be validated (e.g. const or delta const)
		return false
	}
}
```
I want to know why, and are there any cases where the decoding result is inconsistent with the encoded raw data?
I deployed VictoriaMetrics containers in a Kubernetes cluster and reproduced this problem by adding some logs:
```go
// UnmarshalData unmarshals block data.
func (b *Block) UnmarshalData() error {
	/** ... */
	} else if b.bh.TimestampsMarshalType.NeedsValidation() {
		// Ensure timestamps are in the range [MinTimestamp ... MaxTimestamps] and are ordered.
		if err := checkTimestampsBounds(b.timestamps, b.bh.MinTimestamp, b.bh.MaxTimestamp); err != nil {
			fmt.Errorf("[b.UnmarshalData] fail to validate timestamps, err: %v", err)
			fixFlag = true
		}
	}
	/** ... */
	if fixFlag {
		b.fixData()
	}
	/** ... */
}
```
In `(*Block).fixData`, I sorted the out-of-order samples and printed the log:
```
2024-01-25 13:49:25 | 2024-01-25T05:49:25.145Z	warn	VictoriaMetrics/lib/storage/block.go:330	[fixData] fix ended, samples: [{1706161740057 0} {1706161740057 0} {1706161740057 0} {1706161740057 0} {1706161740057 0} {1706161740057 0} {1706161740057 0} {1706161745438 0}], (MinTimestamp, MaxTimestamp): (1706161740057, 1706161745438)
2024-01-25 13:49:25 | 2024-01-25T05:49:25.145Z	warn	VictoriaMetrics/lib/storage/block.go:320	[fixData] fix started, samples: [{1706161745438 0} {1706161740057 0} {1706161740057 0} {1706161740057 0} {1706161740057 0} {1706161740057 0} {1706161740057 0} {1706161740057 0}], (MinTimestamp, MaxTimestamp): (1706161745438, 1706161740057)
2024-01-25 13:49:25 | 2024-01-25T05:49:25.144Z	warn	VictoriaMetrics/lib/storage/block.go:330	[fixData] fix ended, samples: [{1706161736677 0} {1706161736677 0} {1706161736677 0} {1706161736677 0} {1706161736677 0} {1706161736677 0} {1706161736677 0} {1706161745438 0}], (MinTimestamp, MaxTimestamp): (1706161736677, 1706161745438)
2024-01-25 13:49:25 | 2024-01-25T05:49:25.144Z	warn	VictoriaMetrics/lib/storage/block.go:320	[fixData] fix started, samples: [{1706161745438 0} {1706161736677 0} {1706161736677 0} {1706161736677 0} {1706161736677 0} {1706161736677 0} {1706161736677 0} {1706161736677 0}], (MinTimestamp, MaxTimestamp): (1706161745438, 1706161736677)
```
It seems that the decoding results have been corrupted, because:
- they are out of order;
- there are also duplicate samples.

### To Reproduce

It seems random.

### Version

`2fc7e9f47e4e30dbb55860a4eb622de35887bea1`
I compiled and built the image on an `arm64` machine for an `amd64` machine and disabled `CGO` cross-compilation.

**Comment by hagen1778** [comment]
Hello @KofClubs!

> It seems that the decoding results have been corrupted, because:

Do you observe `cannot unmarshal` errors in VM logs during normal operation?

**Comment by KofClubs** [comment]
Yes. According to my observation, these problems occur **randomly** when merging parts:
```
panic: FATAL: unrecoverable error when merging parts in the partition ("/victoria-metrics-data/data/small/2024_01", "/victoria-metrics-data/data/big/2024_01"): cannot merge 14 parts to /victoria-metrics-data/data/small/2024_01/17ACD4BDCF3FFEC1: cannot merge 14 streams: [/victoria-metrics-data/data/small/2024_01/17ACB6B8B993E100 /victoria-metrics-data/data/small/2024_01/17ACD3784F0F200B /victoria-metrics-da..0261,minTimestamp=1705972050993,maxTimestamp=1705972156887}]: cannot unmarshal and calibrate scale for blocks to be merged: timestamp for the row 1 cannot be smaller than the timestamp for the row 0 (total 8 rows); got 1705970270057 vs 1705970281302
```

**Comment by f41gh7** [comment]
Hello, this check protects from possible on-disk data corruption. 

 Could you add a bit more context for your setup:
- Is it a cluster?
- which filesystem is used?
- does readWriteMany option configured for kubernetes volumes?
- Any reason to disabled CGO?

 For now, it looks like on-disk data corruption.

**Comment by KofClubs** [comment]
Thanks Reply.

> Hello, this check protects from possible on-disk data corruption.
Are there other instances of this situation?

**A1:** Yes. VictoriaMetrics runs in a Docker container and the data directory is mounted on a disk.
```bash
[root@iZ8vb83gdoccx3r7hn04cjZ ~]# cd /var/lib/kubelet/plugins/kubernetes.io/csi/pv/d-8vb9pktjuryrzyihhefq/globalmount/
[root@iZ8vb83gdoccx3r7hn04cjZ globalmount]# ll
total 36
drwxr-xr-x 4 root root  4096 Jan 29 14:41 data
-rw-r--r-- 1 root root     0 Jan 29 14:41 flock.lock
drwxr-xr-x 6 root root  4096 Jan 29 14:41 indexdb
drwx------ 2 root root 16384 Jan 29 14:41 lost+found
drwxr-xr-x 2 root root  4096 Jan 29 14:41 metadata
drwxr-xr-x 2 root root  4096 Jan 29 14:41 snapshots
drwxr-xr-x 3 root root  4096 Jan 29 14:41 tmp
```
**A2:** `ext4`. It is an [ESSD](https://www.alibabacloud.com/help/en/ecs/user-guide/essds).
**A3:** No. `accessModes` is `ReadWriteOnce`.
**A4:** I develop on a darwin/arm64 machine, pack the build product into a Docker image, and run it on a linux/amd64 server. The compilation is based on `app-via-docker-linux-amd64`. I found that C/C++-Go cross-compiling and Docker image cross-platform building cannot be executed at the same time, so I disabled the former.

**Comment by valyala** [comment]
VictoriaMetrics uses https://github.com/klauspost/compress library for data compression / decompression when GCO is disabled. The commit 2fc7e9f47e4e30dbb55860a4eb622de35887bea1 uses `v1.17.0` release of this library (see `go.mod` file contents in this commit), while the [latest available release](https://github.com/klauspost/compress/releases) is `v1.17.6`. Probably, there were some bugs in `v1.17.0`, which were fixed in `v1.17.6`.

@KofClubs , could you build VictoriaMetrics from the commit 9c9021fd8b3c95d885c465bbfbe7410aaaf6a70e and verify whether it leads to the `FATAL: unrecoverable error when merging parts in the partition` panic? Note that already existing data stored at `-storageDataPath` may be already corrupted, so it is recommended starting with the empty `-storageDataPath` directory for the new build.

**Comment by KofClubs** [comment]
> @KofClubs , could you build VictoriaMetrics from the commit https://github.com/VictoriaMetrics/VictoriaMetrics/commit/9c9021fd8b3c95d885c465bbfbe7410aaaf6a70e and verify whether it leads to the FATAL: unrecoverable error when merging parts in the partition panic?
I created a new testing instance and observed that this problem has not occurred again. thank you all!
